blog/_posts/2023-12-08-migration-to-scala-3.md

```diff
@@ -0,0 +1,146 @@
+
+There is no one way to do the migration. Decide what’s possible and worthwhile to you (and your company).
+
+Should you rewrite one service at a time? In what order? Or should you leave the old services untouched and write all the new services in Scala 3? What about internal libraries? How much time and resources can you allocate? (The last one might be the hardest to address)
+
+It’s also an excellent time to discuss the benefits and trade-offs. And keep it fair. For example, if you just want a new syntax, is it worth disturbing an old service that hasn’t been touched in years? Well, it’s up to you to decide.
+
+*We decided to write all the new services in Scala 3 (where most of the future work will happen) and keep the rest. Consequently, onboarding new devs without prior Scala experience is one of my biggest challenges.* W*hen they switch from Scala 3 service to Scala 2, they bump into hurdles — some of the “intuitive” concepts don’t work as expected anymore. This requires more explanation; they seem to have to learn some things twice.*
+
+Okay, we defined what migration means, but how exactly do we go about it?
+
+## Step 1. Analyze and plan
+
+Even if we decide not to touch the existing services, we need to review our codebase:
+
+- Do we have any shared internal libraries? Would we need to use them in new services?
+- What external libraries do we use? Do they have a Scala3-compatible release?
+
+Other questions you might need to consider:
+
+- How many services do you need to migrate?
+- What Scala version(s) does each service use?
+
+Thanks to the [interoperability](https://docs.scala-lang.org/scala3/guides/migration/compatibility-source.html) between Scala 3 and Scala 2.13, it might be more accessible or crucial to upgrade to Scala 2.13 first. You have to keep this in mind.
+
+*By this point, we knew that all our services were on 2.13; we knew how many internal libraries needed to be used from both Scala versions and what external libraries had no Scala 3 support: what is the status, what is missing, and what can we do to deal with it. For example, a couple of libraries got a Scala 3 release by the time we were done with the investigation, and one `circe` ”extras” library was replaced with two lines of boilerplate.*
+
+## Step 2. Draw the rest of the owl
+
+After all the meetings and/or writings comes the time to do the actual work. When you have concrete questions or action items, remember that [Scala 3 migration guides](https://docs.scala-lang.org/scala3/guides/migration/compatibility-intro.html) are your friends. For example, it covers how to port compiler options and sbt projects. Great reference resource.
```

Comment:
Also please include the link to the Scala Discord where people can find further help from the community.